File: zenoh-grpc-client-sdk/zenoh-grpc-python/examples/sub_callback.py
import zenoh_grpc
import time

# with zenoh_grpc.Session.connect("unix:///tmp/zenoh-grpc.sock") as session:
# with zenoh_grpc.Session.connect("tcp://127.0.0.1:7335") as session:
with zenoh_grpc.Session.connect() as session:
    def on_sample(sample):
        print("\ncallback:", sample.key_expr, sample.payload, sample.encoding)
        # print("dropped count:", sub.dropped_count())
        # time.sleep(1) # do not sleep here, just for testing the dropped count

    sub = session.declare_subscriber("demo/example/**", 
                                     callback=on_sample,
                                     allowed_origin=zenoh_grpc.Locality.ANY) # optional, default is ANY
    time.sleep(1000)

File: zenoh-grpc-client-sdk/zenoh-grpc-cpp/examples/sub_callback.cpp
#include <iostream>

#include "common.hpp"

static void on_sample(const zenoh_grpc::SubscriberEvent& event) {
    std::cout << "callback: ";
    if (event.sample.has_value()) {
        print_sample(*event.sample);
    } else {
        std::cout << "<empty event>" << std::endl;
    }
}

int main(int argc, char** argv) {
    auto session = zenoh_grpc::Session::connect(example_endpoint(argc, argv));

    zenoh_grpc::Session::SubscriberOptions options;
    options.allowed_origin = zenoh_grpc::Locality::Any;
    auto subscriber = session.declare_subscriber("demo/example/**", on_sample, options);

    std::cout << "callback subscriber ready on " << example_endpoint(argc, argv) << std::endl;
    example_sleep_ms(5000000);
    (void)subscriber;
    return 0;
}
